src/rsdp_search.rs

```diff
@@ -0,0 +1,126 @@
+use core::{mem, ops::RangeInclusive};
+use rsdp::Rsdp;
+use {parse_validated_rsdp, AcpiError, AcpiHandler};
+
+/// The pointer to the EBDA (Extended Bios Data Area) start segment pointer
+const EBDA_START_SEGMENT_PTR: usize = 0x40e;
+/// The earliest (lowest) memory address an EBDA (Extended Bios Data Area) can start
+const EBDA_EARLIEST_START: usize = 0x80000;
+/// The end of the EBDA (Extended Bios Data Area)
+const EBDA_END: usize = 0x9ffff;
+/// The start of the main bios area below 1mb in which to search for the RSDP
+/// (Root System Description Pointer)
+const RSDP_BIOS_AREA_START: usize = 0xe0000;
+/// The end of the main bios area below 1mb in which to search for the RSDP
+/// (Root System Description Pointer)
+const RSDP_BIOS_AREA_END: usize = 0xfffff;
+/// The RSDP (Root System Description Pointer)'s signature, "RSD PTR " (note trailing space)
+const RSDP_SIGNATURE: &'static [u8; 8] = b"RSD PTR ";
+
+/// Find the begining of the EBDA (Extended Bios Data Area) and return `None` if the ptr at
+/// `0x40e` is invalid.
+pub fn find_search_areas<H>(handler: &mut H) -> [RangeInclusive<usize>; 2]
+where
+    H: AcpiHandler,
+{
+    // Read base segment from BIOS area. This is not always given by the bios, so it needs to be
+    // checked. We left shift 4 because it is a segment ptr.
+    let base_mapping =
+        handler.map_physical_region::<u16>(EBDA_START_SEGMENT_PTR, mem::size_of::<u16>());
+    let base = (*base_mapping as usize) << 4;
+    handler.unmap_physical_region(base_mapping);
+
+    // Check if base segment ptr is in valid range valid
+    let ebda_start = if (EBDA_EARLIEST_START..EBDA_END).contains(&base) {
+        debug!("EBDA address is {:#x}", base);
+        Some(base)
+    } else {
+        warn!(
+            "EBDA address at {:#x} out of range ({:#x}), falling back to {:#x}",
+            EBDA_START_SEGMENT_PTR, base, EBDA_EARLIEST_START
+        );
+
+        None
+    };
+
+    [
+        // Main bios area below 1 mb
+        // In practice (from my [Restioson's] testing, at least), the RSDP is more often here than
+        // the in EBDA. Also, if we cannot find the EBDA, then we don't want to search the largest
+        // possible EBDA first.
+        RSDP_BIOS_AREA_START..=RSDP_BIOS_AREA_END,
+        if let Some(ebda_start) = ebda_start {
```

Comment:
This could then be written:
``` rust
if (EBDA_EARLIEST_START..EBDA_END).contains(&ebda_start) {
    ebda_start..=ebda_start + 1024
} else {
     // comment
    EBDA_EARLIEST_START..=EBDA_END
}
```